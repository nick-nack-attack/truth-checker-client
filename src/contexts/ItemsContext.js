// Contexts for all the content
import React, {
    createContext,
    useReducer
} from "react";

let ItemsContext = createContext();

let initialState = {
    fetched: false,
    facts: [],
    reports: [],
    all: [],
    error: null
};

// Allows dispatch actions to update state
let reducer = (state, action) => {
    switch (action.type) {
        // set facts from the response
        case 'set-items':
            return {
                ...state,
                fetched: true,
                facts: action.payload.facts,
                reports: action.payload.reports,
                all: action.payload.all
            }
            // update only the facts
            case 'set-facts':
                return {
                    ...state,
                    facts: action.payload,
                }
            case 'set-reports':
                return {
                    ...state,
                    reports: action.payload,
                    fetched: true
                }
            // set all items
            case 'set-all':
                return {
                    ...state,
                    all: action.payload
                }
            // refetch items from database
            case 'refetch':
                return {
                    ...state,
                    fetched: false
                }
            // set error in state
            case 'set-error':
                return {
                    ...state,
                    error: action.payload
                }
            default:
                return initialState
    }
};

const ItemsContextProvider = (props) => {
    let [ state, dispatch ] = useReducer(reducer, initialState);
    let value = { state, dispatch };
    return (
        <ItemsContext.Provider value={value}>
            { props.children }
        </ItemsContext.Provider>
    );
};

let ItemsContextConsumer = ItemsContext.Consumer;

export { ItemsContext, ItemsContextProvider, ItemsContextConsumer };