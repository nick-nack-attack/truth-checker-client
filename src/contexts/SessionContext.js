import React, {
    createContext,
    useReducer
} from "react";

// utilities

let SessionContext = createContext();

let initialState = {
    menu: {
        open: false,
        toggleMenu: () => {}
    }
};

// allow dispatch actions to update state
let reducer = (state, action) => {
    switch (action.type) {
        // start session
        case "start-session":
            return {
                ...state,
            }

        // open or close menu
        case "toggle-menu":
            return {
                ...state,
                menu: {
                    open: !state.menu.open
                }
            }

        default:
            return {
                ...initialState
            };
    }
};

const SessionContextProvider = (props) => {
    let [ state, dispatch ] = useReducer(reducer, initialState);
    let value = { state, dispatch };
    return (
        <SessionContext.Provider value={value}>
            { props.children }
        </SessionContext.Provider>
    );
};

let SessionContextConsumer = SessionContext.Consumer;

export { SessionContext, SessionContextProvider, SessionContextConsumer };