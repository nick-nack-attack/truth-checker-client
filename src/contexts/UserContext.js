// context for the user
import React, {
    createContext,
    useReducer
} from "react";

let UserContext = createContext();

let initialState = {
    name: 'Admin',
    isLoggedIn: false,
    fetched: false
};

let reducer = (state, action) => {
    switch (action.type) {
        // logs user in
        case "login":
            return {
                isLoggedIn: true,
                fetched: true
            };
        case "logout":
            return {
                // logs user out
                name: '',
                isLoggedIn: '',
                fetched: false
            }
        default:
            return initialState
    }
};

const UserContextProvider = props => {
    let [ state, dispatch ] = useReducer(reducer, initialState);
    let value = { state, dispatch };
    return (
        <UserContext.Provider value={value}>
            { props.children }
        </UserContext.Provider>
    );
};

let UserContextConsumer = UserContext.Consumer;

export { UserContext, UserContextProvider, UserContextConsumer };
