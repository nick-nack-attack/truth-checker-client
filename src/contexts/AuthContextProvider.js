// this file has all the contexts needs in the app
import React from 'react';

// here are the context providers
import {ItemsContextProvider} from './ItemsContext';
import {SessionContextProvider} from './SessionContext';

const AuthContextProvider = (props) => {
  return (
      <SessionContextProvider>
        <ItemsContextProvider>
          {props.children}
        </ItemsContextProvider>
      </SessionContextProvider>
  )
};

export default AuthContextProvider;