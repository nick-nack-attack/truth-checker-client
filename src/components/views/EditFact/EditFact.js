// admin can edit a fact on this page
import React, {useContext, useState} from 'react';
import PropTypes from 'prop-types';
import {useHistory} from 'react-router-dom';

// contexts and hooks
import {ItemsContext} from '../../../contexts/ItemsContext';
import {UseInputChange} from '../../../hooks/UseInputChange';

// helpers
import config from '../../../config';

// components
import Error from '../../utils/Error/Error';
import Form from '../../utils/Form/Form';

// service
import FactsApiService from '../../../services/facts-service';

const EditFact = props => {

  // initialize variables
  const itemsContext = useContext(ItemsContext);
  const history = useHistory();
  const [input, handleInputChange] = UseInputChange({});
  const [errors, setErrors] = useState({});
  const [fact,] = useState(props.fact);

  const validateEditFactForm = (e) => {
    let errors = {};
    e.preventDefault();
    const numOfValues = Object.values(input).filter(Boolean).length;
    if (numOfValues === 0) {
      errors.title = {message: 'Must change at least one value to update'};
      setErrors(errors);
    } else {
      setErrors({});
      submitForm();
    }
  };

  const submitForm = () => {
    const fact_id = fact.fact_id;
    const fields = {
      ...input,
      // public user is hardcoded to 2
      user_id: 2
    };
    FactsApiService.updateFact(fact_id, fields)
        .then(() => {
          itemsContext.dispatch({
            type: 'refetch'
          });
          // on success, pass up to notification
          // and push to fact feed
          props.onSuccess('edit-fact');
          history.push(config.FACTS_FEED);
        })
  };

  const handleClickDelete = (id, e) => {
    e.preventDefault();
    if (window.confirm(`Are you sure you want to delete this Fact?`)) {
      FactsApiService.deleteFact(id)
          .then(() => {
            itemsContext.dispatch({
              type: 'refetch'
            });
            props.onSuccess('delete-fact');
            history.push(config.FACTS_FEED);
          });
    }
  };

  // return form
  return (
      <>
        <Form
            formType="Edit-Fact"
            validateEditFactForm={validateEditFactForm}
            handleInputChange={handleInputChange}
            title={fact.title}
            submitted={fact.date_submitted}
            underReview={fact.date_under_review}
            approved={fact.date_approved}
            notTrue={fact.date_not_true}
            handleCancelClick={() => history.push(config.FACTS_FEED)}
            handleClickDelete={e => handleClickDelete(fact.fact_id, e)}
        />
        {errors.title
            ? <Error
                message={errors.title.message}
            />
            : ""
        }
      </>
  )

};

EditFact.propTypes = {
  title: PropTypes.string,
  fact_id: PropTypes.number,
  date_submitted: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  date_under_review: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  date_approved: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  date_not_true: PropTypes.oneOfType([PropTypes.string, PropTypes.object])
};

export default EditFact;