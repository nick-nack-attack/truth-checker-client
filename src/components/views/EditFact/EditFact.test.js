import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json'
import { BrowserRouter as Router } from 'react-router-dom';
import EditFact from './EditFact';

import { UserContextProvider } from '../../../contexts/UserContext'
import AuthContextProvider from '../../../contexts/AuthContextProvider';

const context = React.createContext({
  rmbrArray: []
});

export class testContext extends Component {
  state = {
      rmbrArray: ['One item']
  }
  render() {
      const value = {
          rmbrArray: this.state.rmbrArray
      }
      return (
          <testContext.Provider value={value}>
              { this.props.children }
          </testContext.Provider>
      )
  }
}

describe('Edit Fact Component', () => {

  it('renders without crashing', () => {
    let fact_id = 1;
    const wrapper = shallow(
      <AuthContextProvider>
        <testContext>
          <EditFact 
            fact_id={fact_id}
            context={context}
          />
        </testContext>
      </AuthContextProvider>
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })

});