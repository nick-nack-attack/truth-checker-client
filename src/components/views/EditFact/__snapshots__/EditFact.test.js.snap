// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Fact Component renders without crashing 1`] = `
<SessionContextProvider>
  <ItemsContextProvider>
    <testContext>
      <EditFact
        context={
          Object {
            "$$typeof": Symbol(react.context),
            "Consumer": Object {
              "$$typeof": Symbol(react.context),
              "_calculateChangedBits": null,
              "_context": [Circular],
            },
            "Provider": Object {
              "$$typeof": Symbol(react.provider),
              "_context": [Circular],
            },
            "_calculateChangedBits": null,
            "_currentRenderer": null,
            "_currentRenderer2": null,
            "_currentValue": Object {
              "rmbrArray": Array [],
            },
            "_currentValue2": Object {
              "rmbrArray": Array [],
            },
            "_threadCount": 0,
          }
        }
        fact_id={1}
      />
    </testContext>
  </ItemsContextProvider>
</SessionContextProvider>
`;
